File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
       
    
    int rob(vector<int>& nums) {
        
        if(nums.size()==1) return nums[0];
        int n=nums.size();
        vector<int>mem(nums.size(),0);
        mem[0]=nums[0];
        mem[1]=nums[1];
        for(int i=2;i<n;i++){
                mem[i]=max(mem[i-1]-nums[i-1],mem[i-2])+nums[i]; 
        }
        
        
        
        return max(mem[n-1],mem[n-2]);
        
    }
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    
    def rob(self, nums: List[int]) -> int:
        
            dp=[0]*(len(nums)+2)
            
            
            for i in range(len(nums)):
                
                val=i+2
                dp[val]=max(dp[val-1],dp[val-2]+nums[i])
                
                
            
            return max(dp[-2:])